library/core/src/str/lossy.rs

```diff
@@ -115,34 +103,39 @@ impl<'a> Iterator for Utf8LossyChunksIter<'a> {
                             (0xF0, 0x90..=0xBF) => (),
                             (0xF1..=0xF3, 0x80..=0xBF) => (),
                             (0xF4, 0x80..=0x8F) => (),
-                            _ => {
-                                error!();
-                            }
+                            _ => break,
                         }
                         i += 1;
                         if safe_get(self.source, i) & 192 != TAG_CONT_U8 {
-                            error!();
+                            break;
                         }
                         i += 1;
                         if safe_get(self.source, i) & 192 != TAG_CONT_U8 {
-                            error!();
+                            break;
                         }
                         i += 1;
                     }
-                    _ => {
-                        error!();
-                    }
+                    _ => break,
                 }
             }
+
+            valid_up_to = i;
         }
 
-        let r = Utf8LossyChunk {
-            // SAFETY: We have checked that the entire source is valid UTF-8.
-            valid: unsafe { from_utf8_unchecked(self.source) },
-            broken: &[],
-        };
-        self.source = &[];
-        Some(r)
+        // SAFETY: `i <= self.source.len()` because it only ever increments by 1
```

Comment:
`i` is incremented up to 3 times in some loop iterations?

Reply:
It's incremented up to 4 times in some loop iterations, but what I meant is: in between every pair of `i += 1`, we always compare `i` vs `self.source.len()`. As soon as an `i += 1` puts `i` out of bounds, then `if safe_get(self.source, i) & 192 != TAG_CONT_U8 { break; }` will terminate the loop because 0 &amp; 192 != 128.